### PR TITLE
Feature/automation

### PR DESCRIPTION
**Pull request by NayanaBannur**

### CODE REVIEW

File: covid19-library/src/data_fetchers/data_fetcher_base.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+from entities.forecast_variables import ForecastVariable
+from data_fetchers.data_fetcher_utils import load_regional_metadata
+
+ABC = abc.ABCMeta('ABC', (object,), {'__slots__': ()})
+
+class DataFetcherBase(ABC):
+    @property
+    @abc.abstractmethod
+    def get_observations_for_region_single(self, region_type: str, region_name:str):
```
Comment (anupama-agarwal) [inline covid19-library/src/data_fetchers/data_fetcher_base.py]: rename to get_observations_for_single_region

File: covid19-library/src/data_fetchers/data_fetcher_base.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        pass
+
+    def get_regional_metadata_single(self, region_type: str, region_name: str, filepath: str):
```
Comment (anupama-agarwal) [inline covid19-library/src/data_fetchers/data_fetcher_base.py]: same here

File: covid19-library/src/data_fetchers/data_fetcher_base.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        metadata = load_regional_metadata(filepath)
+        for params in metadata["regional_metadata"]:
+            if params["region_type"] == region_type and params["region_name"] == region_name:
+                return params["metadata"]
+
+
+    def get_observations_for_region(self, region_type: str, region_name:str, smooth: bool = True):
```
Comment (anupama-agarwal) [inline covid19-library/src/data_fetchers/data_fetcher_base.py]: region_name: list[str]..
maybe call it region_names

File: covid19-library/src/data_fetchers/data_fetcher_base.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        combined_region_name = " ".join(region_name)
+
+        if len(region_name) > 1:
+            cols = [col for col in list(df) if col not in {"region_name","region_type","observation"}]
+            df = df.groupby(["observation"])[cols].sum().reset_index()
+            df.insert(0, column="region_name", value=combined_region_name)
+            df.insert(1, column="region_type", value=region_type)
+
+        if smooth:
+            window_size, min_window_size = 3, 1
```
Comment (anupama-agarwal) [inline covid19-library/src/data_fetchers/data_fetcher_base.py:45]: we can move smoothening logic to separate function

File: covid19-library/src/data_fetchers/data_fetcher_base.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    def get_regional_metadata(self, region_type: str, region_name: str, filepath: str):
+        """
+        region_type : Type of region (city, district, state)
+        region_name : List of regions
+        filepath : Path to file containing regional metadata
+        """
+        population = 0
+        for region in region_name:
+            regional_metadata = self.get_regional_metadata_single(region_type, region, filepath)
+            population += regional_metadata["population"]
```
Comment (anupama-agarwal) [inline covid19-library/src/data_fetchers/data_fetcher_base.py]: iterate over keys and do addition.

Also add a comment saying we are not sure if metadata will always be additive.

File: covid19-library/src/data_fetchers/official_data.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    df_result.insert(0, column = "region_name", value = district.lower().replace(',', ''))
+    df_result.insert(1, column = "region_type", value = "district")
+
+    return df_result
+
+
+class OfficialData(DataFetcherBase):
+
+    def get_observations_for_region_single(self, region_type, region_name):
+        return get_data_from_db(region_name)
```
Comment (anupama-agarwal) [inline covid19-library/src/data_fetchers/official_data.py:108]: if region_type is not a district throw a not implemented error.

File: covid19-library/src/data_fetchers/tracker_raw.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+import urllib.request
+from datetime import datetime
+from io import StringIO  ## for Python 3
+from functools import lru_cache
+
+# raw input data from cocvid19india.org
+raw_data_url = "https://api.covid19india.org/raw_data.json"
+raw_data_urls = ["https://api.covid19india.org/raw_data1.json", "https://api.covid19india.org/raw_data2.json"]
+post_april27_url = "https://api.covid19india.org/raw_data3.json"
+
```
Comment (anupama-agarwal) [inline covid19-library/src/data_fetchers/tracker_raw.py:17]: we need to additional raw_data urls

File: covid19-library/src/modules/data_fetcher_module.py [hunk 8]
```diff
@@ -1,153 +1,38 @@
...
+    def get_regional_metadata(region_type, region_name, 
+    data_source = DataSource.tracker_district_daily, filepath="../data/regional_metadata.json"):
+        data_fetcher = DataFetcherFactory.get_data_fetcher(data_source)
+        metadata = data_fetcher.get_regional_metadata(region_type, region_name, filepath)
+        return metadata
 
 
 if __name__ == "__main__":
-    DataFetcherModule.get_observations_for_region("district", "bengaluru")
+    print(DataFetcherModule.get_observations_for_region(
```
Comment (anupama-agarwal) [inline covid19-library/src/modules/data_fetcher_module.py]: we can remove this.

File: covid19-library/src/modules/forecasting_module.py [hunk 9]
```diff
@@ -48,21 +48,20 @@ def convert_to_old_required_format(self, run_day, predictions_df, region_type, r
...
         preddf.insert(4, 'Error', "MAPE")
         preddf.insert(5, "Error Value", error * 100)
 
         return preddf
 
-    def predict_for_region(self, region_type, region_name, run_day, forecast_start_date,
-                           forecast_end_date, add_initial_observation):
-        observations = DataFetcherModule.get_observations_for_region(region_type, region_name)
-        region_metadata = DataFetcherModule.get_regional_metadata(region_type, region_name)
+    def predict_for_region(self, data_source, region_type, region_name, run_day, forecast_start_date,
```
Comment (anupama-agarwal) [inline covid19-library/src/modules/forecasting_module.py:58]: we can put a default data source.

File: covid19-library/src/modules/training_module.py [hunk 10]
```diff
@@ -51,17 +51,17 @@ def optimize(self, search_space, region_metadata, region_observations, train_sta
         metrics_result = ModelEvaluator.evaluate_for_forecast(region_observations, predict_df, [loss_function])
         return metrics_result[0]["value"]
 
-    def train_for_region(self, region_type, region_name, train_start_date, train_end_date,
+    def train_for_region(self, data_source, region_type, region_name, train_start_date, train_end_date,
```
Comment (anupama-agarwal) [inline covid19-library/src/modules/training_module.py:54]: default for data_source is required.

File: covid19-library/src/configs/base_config.py [hunk 11]
```diff
@@ -8,7 +8,8 @@
 
 
 class BaseConfig(BaseModel):
-    region_name: str
+    data_source: str
```
Comment (anupama-agarwal) [inline covid19-library/src/configs/base_config.py]: put a default here.

**Review by anupama-agarwal — CHANGES_REQUESTED** [review]

File: covid19-library/src/data_fetchers/tracker_raw.py (hunk 7 above)
Comment (NayanaBannur) [reply covid19-library/src/data_fetchers/tracker_raw.py:17]: To be done next.

File: covid19-library/src/data_fetchers/data_fetcher_base.py (hunk 4 above)
Comment (NayanaBannur) [reply covid19-library/src/data_fetchers/data_fetcher_base.py:45]: To be done next.

**Pull request merged**